flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java

```diff
@@ -0,0 +1,242 @@
+				indexBuffer.putInt(subpartitionBuffers[subpartition]);
+				totalBytesWritten += subpartitionBytes[subpartition];
+			}
+
+			++numRegions;
+			currentSubpartition = 0;
+			Arrays.fill(subpartitionBytes, 0);
+			Arrays.fill(subpartitionBuffers, 0);
+		}
+	}
+
+	private void flushIndexBuffer() throws IOException {
+		indexBuffer.flip();
+		if (indexBuffer.limit() > 0) {
+			BufferReaderWriterUtil.writeBuffer(indexFileChannel, indexBuffer);
+		}
+	}
+
+	/**
+	 * Writes a {@link Buffer} of the given subpartition to the this {@link PartitionedFile}.
+	 *
+	 * <p>Note: The caller is responsible for recycling the target buffer and releasing the failed
+	 * {@link PartitionedFile} if any exception occurs.
+	 */
+	public void writeBuffer(Buffer target, int targetSubpartition) throws IOException {
+		checkArgument(targetSubpartition >= currentSubpartition, "Must write in subpartition index order.");
+		checkState(!isFinished, "File writer is already finished.");
+		checkState(!isClosed, "File writer is already closed.");
+		checkState(dataFileChannel != null && indexFileChannel != null, "Must open the partitioned file first.");
+
+		currentSubpartition = Math.max(currentSubpartition, targetSubpartition);
+		long numBytes = BufferReaderWriterUtil.writeToByteChannel(dataFileChannel, target, header);
+
+		++subpartitionBuffers[targetSubpartition];
+		subpartitionBytes[targetSubpartition] += numBytes;
+	}
+
+	/**
+	 * Finishes writing which closes the file channel and returns the corresponding {@link PartitionedFile}.
+	 *
+	 * <p>Note: The caller is responsible for releasing the failed {@link PartitionedFile} if any exception
+	 * occurs.
+	 */
+	public PartitionedFile finish() throws IOException {
+		checkState(!isFinished, "File writer is already finished.");
+		checkState(!isClosed, "File writer is already closed.");
+		checkState(dataFileChannel != null && indexFileChannel != null, "Must open the partitioned file first.");
+
+		isFinished = true;
+
+		writeRegionIndex();
+		flushIndexBuffer();
+		indexBuffer.rewind();
+
+		close();
+		ByteBuffer indexDataCache = allIndexDataCached ? indexBuffer : null;
+		return new PartitionedFile(numRegions, numSubpartitions, dataFilePath, indexFilePath, indexDataCache);
+	}
+
+	/**
+	 * Used to close and delete the failed {@link PartitionedFile} when any exception occurs.
+	 */
+	public void releaseQuietly() {
+		IOUtils.closeQuietly(this);
+		IOUtils.deleteFileQuietly(dataFilePath);
+		IOUtils.deleteFileQuietly(indexFilePath);
+	}
+
+	@Override
+	public void close() {
+		if (isClosed) {
+			return;
+		}
+		isClosed = true;
+
+		Throwable exception = null;
+		try {
+			dataFileChannel.close();
+		} catch (Throwable throwable) {
+			exception = throwable;
+		}
+
+		try {
+			indexFileChannel.close();
+		} catch (Throwable throwable) {
+			exception = ExceptionUtils.firstOrSuppressed(throwable, exception);
+		}
+
+		if (exception != null) {
+			ExceptionUtils.rethrow(exception);
```

Comment:
Nit: You can add `throws Exception` to the method signature (compatible with `AutoClosable`) and then you don't need to wrap it as a RuntimeException.

Reply:
I use ExceptionUtils.rethrow for it can handle both Error and Exception.

Reply:
In this case it does a redundant wrapping into RuntimeException. If you add the exception signature you avoid that.
Exception wrapping should be avoided unless you add actual information in the new exception. Otherwise it just complicates the stack traces for debugging.